monai/inferers/splitter.py

```diff
@@ -58,142 +92,369 @@ def __call__(self, inputs: Any) -> Iterable[tuple[torch.Tensor, Sequence[int]]]:
-                    f"`patch_filter_fn` can have at most two positional parameters (patch, location)."
-                    f"The provided callable ({filter_fn}) has {n_pos_params} positional parameters."
-                )
-            return filter_fn
-        raise ValueError(
-            "`patch_filter_fn` should be a callable with two input parameters (patch, location). "
-            f"{type(filter_fn)} is given."
-        )
-
-    def _get_valid_patch_size(self, spatial_ndim):
-        return ensure_tuple_rep(self.patch_size, spatial_ndim)
-
-    def _get_valid_overlap(self, spatial_ndim, patch_size):
-        # broadcast overlap is possible
-        overlap = ensure_tuple_rep(self.overlap, spatial_ndim)
-        # keep overlap only in patching dimensions
-        return tuple(o if p else 0.0 for o, p in zip(overlap, patch_size))
-
-    def _get_valid_offset(self, spatial_shape, spatial_ndim, patch_size):
-        offset = ensure_tuple_rep(self.offset, spatial_ndim)
-        for off, ps, ins in zip(offset, patch_size, spatial_shape):
-            if off < 0 and not self.pad_mode:
+            elif num_pos_params > 2:
                 raise ValueError(
-                    f"Negative `offset` ({off}) requires a valid padding mode, "
-                    f"but `pad_mod` is set to {self.pad_mode}."
+                    f"`filter_fn` can have at most two positional parameters (patch, location)."
+                    f"The provided callable ({filter_fn}) has {num_pos_params} positional parameters."
                 )
-            if off < -ps:
-                raise ValueError(f"Negative `offset` ({off}) cannot be larger than `patch_size` ({ps}) in magnitude.")
-            if off >= ins:
-                raise ValueError(f"`offset` ({off}) cannot be larger than inputs size ({ins}).")
-        return offset
+        elif filter_fn is not None:
+            raise ValueError(
+                "`filter_fn` should be a callable with two input parameters (patch, location). "
+                f"{type(filter_fn)} is given."
+            )
+        return filter_fn
 
     def _calculate_pad_size(self, spatial_shape, spatial_ndim, patch_size, offset, overlap):
-        if not self.pad_mode:
-            return [], False
         # initialize with zero
         pad_size = [0] * 2 * spatial_ndim
+        if not self.pad_mode:
+            return pad_size, False
         # set the starting pad size only if the offset is negative
         pad_size[1::2] = (-min(off, 0) for off in offset)
         # set the ending pad size only if it is not divisible by the patch size
         pad_size[::2] = (
-            0 if ps == 0 else (off - ins + ps) % round(ps * (1.0 - ov))
-            for ins, off, ps, ov in zip(spatial_shape, offset, patch_size, overlap)
+            0 if ps == 0 else (off - sh + ps) % round(ps - (ps * ov if isinstance(ov, float) else ov))
+            for sh, off, ps, ov in zip(spatial_shape, offset, patch_size, overlap)
         )
```

Comment:
This expression is so complex now it might be better as a for loop with `if-else` statements rather than expressions.